internal/utils/time_interval.go

```diff
@@ -0,0 +1,130 @@
+package utils
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+)
+
+const (
+	// ErrEndBeforeStart is an the error message for when a TimeInterval's end time
+	// would be before its start.
+	ErrEndBeforeStart = "end time before start time"
+
+	errWindowTooLargeFmt = "random window equal to or larger than TimeInterval: window %v, interval %v"
+)
+
+// TimeInterval represents an interval of time in UTC. That is, regardless of
+// what timezone(s) are used for the beginning and end times, they will be
+// converted to UTC and methods will return them as such.
+type TimeInterval struct {
+	start time.Time
+	end   time.Time
+}
+
+// NewTimeInterval creates a new TimeInterval for a given start and end. If end
+// is a time.Time before start, then an error is returned.
+func NewTimeInterval(start, end time.Time) (*TimeInterval, error) {
+	if end.Before(start) {
+		return nil, fmt.Errorf(ErrEndBeforeStart)
+	}
+	return &TimeInterval{start.UTC(), end.UTC()}, nil
+}
+
+// Duration returns the time.Duration of the TimeInterval.
+func (ti *TimeInterval) Duration() time.Duration {
+	return ti.end.Sub(ti.start)
+}
+
+// Overlap detects whether the given TimeInterval overlaps with this
+// TimeInterval, assuming an inclusive start boundary and exclusive end
+// boundary.
+func (ti *TimeInterval) Overlap(other *TimeInterval) bool {
```

Comment:
Is this function missing a case? When the first interval starts before second one and ends after the second interval and vice-versa.

I feel this could be written simpler, exclude all the cases where there is no overlap first, since there are less of those. Everything else is overlap.

For example, keep the first check for opposite boundaries and do another check for start and end of this before start of that  and the last check for start and end of this after end of that.

Reply:
After reading the tests, I'm probably wrong on the first part of the comment. But I guess it just solidifies the second part and the comment under this one.

Reply:
The first part you mention is covered by the 3rd case -- first starts before the second, so we need to make sure the second one starts before the first one ends. Maybe it was a readability issue, so it's easier with clearer names.

But I like your solution better, it is cleaner/simpler. So I went with that.